### PR TITLE
[FrontEnd] Add Model Tuner front-end tool.

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary**
Add a front-end tool for tuning (calibrating) the quantization parameters of a model. The parameters are tuned using the model accuracy as the optimization metric.

**Motivation**
The motivation for this tool is this: when using extremely brutal quantization schemes (for example the "SymmetricWithPower2Scale") the accuracy difference between the floating-point model and the quantized model can be pretty high (up to 10's of percents). One such example is an internally designed model which has:
- 99 % accuracy for the floating-point model
- 81 % accuracy for the  (initial) quantized model using SymmetricWithPower2Scale schema
- 98 % accuracy for the (final) quantized model after tuning

Attached here is the console output of this Tool after running it for the model:
- [ModelTuner_Log.txt](https://github.com/pytorch/glow/files/3886201/ModelTuner_Log.txt)

**Algorithm**
The quantization parameters are initially chosen such that no saturation occurs (quantized range includes the min/max of the profile). For some of the tensors where the histogram exhibits some outlier values it might be better to use quantization parameters which saturate the outliers with the benefit of having smaller quantization step for the bulk of the histogram.
The tuning algorithm "plays" with the **scale** quantization parameter by trying sequentially for each node new values around the original value and picks the one which provides the best accuracy (it tries original_scale, original_scale/2 and original_scale/4). One might say that there is a philosophical problem with this approach: the algorithm is over-fitting the quantization parameters for a given dataset. But from a practical point of view it is better to have maybe an over-fitting of a couple of percent than have an under-fitting of 10s of percent for the quantization mechanism.

**Refactorizations**

- Had to refactor a a little the Loader class in order to provide more information through its API for the tools which is using the Loader.
- Refactored Base.cpp to provide a function for validating the quantization parameters.
- Refactored ProtobufLoader.cpp to add function to retrieve the unique input placeholder for a model with a single input placeholder.

**Documentation**
doc/ModelTuner.md

**Test Plan**
None

### CODE REVIEW

**Comment by mciprian13** [comment]
Question: what is the model-runner front-end application intended for? It assumes the graph inputs are constants so I`m not sure what is the purpose of that.

File: tools/loader/LoaderUtils.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+void glow::getDuration(TimeStamp &startTime, unsigned &sec, unsigned &min,
+                       unsigned &hrs) {
+  unsigned totSec = glow::getDurationSec(startTime);
+  unsigned totMin = totSec / 60;
+  sec = totSec % 60;
+  min = totMin % 60;
+  hrs = totMin / 60;
+}
```
Comment (jfix71) [inline tools/loader/LoaderUtils.cpp]: nit: new line at end of file.

File: include/glow/Importer/ProtobufLoader.h [hunk 2]
```diff
@@ -195,10 +195,23 @@ class ProtobufLoader {
     return outputVarsByName_.begin()->second;
   }
 
+  /// \returns the single input of the network. The function assumes that there
+  /// is only one input, returns Error otherwise. For most of the models the
+  /// single input is usually an image tensor.
+  Expected<Placeholder *> getSingleInput() {
+    RETURN_ERR_IF_NOT(inputVarsByName_.size() == 1,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply include/glow/Importer/ProtobufLoader.h]: I just mirrored the already existing function "getSingleOutput". Should I put that also to the cpp?

File: tools/loader/CMakeLists.txt [hunk 3]
```diff
@@ -82,3 +82,21 @@ target_link_libraries(model-compiler
...
+                        Backends
+                        Base
+                        Converter
+                        Graph
+                        HostManager
+                        Importer
+                        ExecutionEngine
+                        GraphOptimizer
+                        Quantization
+                        LLVMSupport)
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tools/loader/CMakeLists.txt]: Done.

File: tools/loader/ImageClassifier.cpp [hunk 4]
```diff
@@ -252,14 +252,13 @@ buildAndCompileAndGetInAndOutPair(Loader &loader, PlaceholderBindings &bindings,
...
+  cctx.bindings = &bindings;
   cctx.backendOpts.autoInstrument = autoInstrument;
   loader.compile(cctx);
 
-  // The image name that the model expects must be passed on the command line.
-  const char *inputName = modelInputName.c_str();
-  Placeholder *inputImagePH =
-      llvm::cast<Placeholder>(EXIT_ON_ERR(LD->getNodeValueByName(inputName)));
+  // Get the input image placeholder.
+  Placeholder *inputImagePH = EXIT_ON_ERR(LD->getSingleInput());
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tools/loader/ImageClassifier.cpp]: Reverted it.

File: tools/loader/LoaderUtils.cpp (hunk 1 above)
Comment (mciprian13) [reply tools/loader/LoaderUtils.cpp]: Done.

File: include/glow/Importer/ProtobufLoader.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jfix71) [reply include/glow/Importer/ProtobufLoader.h]: Sure, thanks!

[the remark this answers is not in the document]
Comment (mciprian13) [reply include/glow/Importer/ProtobufLoader.h]: Done.

**Pull request closed without merging**